Deprecate `toggleableActiveColor`

TahaTesser opened the pull request:
Related https://github.com/flutter/flutter/pull/90317 and #91772
Fixes https://github.com/flutter/flutter/issues/93709

This PR replaces`toggleableActiveColor` as it is soon to be deprecated.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -293,58 +296,87 @@ class RadioListTile<T> extends StatelessWidget {
     }
+    final Set<MaterialState> activeStates = states..add(MaterialState.selected);
+    final Color? effectiveActiveColor = widget.activeColor
+      ?? radioThemeData.overlayColor?.resolve(activeStates);
```
If `Theme.of(context).toggleableActiveColor` is left here as the final undocumented default (ever after it's deprecated), we will not break most existing apps. That will give us some time to migrate developers to radioThemeData.overlayColor,

TahaTesser replied:
Great idea!

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -101,7 +104,7 @@ import 'theme_data.dart';
 ///  * [SwitchListTile], a similar widget for switches.
 ///  * [ListTile] and [Radio], the widgets from which this widget is made.
-class RadioListTile<T> extends StatelessWidget {
+class RadioListTile<T> extends StatefulWidget {
```
Why not leave this as a StatelessWidget?

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -293,58 +296,87 @@ class RadioListTile<T> extends StatelessWidget {
     }
+    final Set<MaterialState> activeStates = states..add(MaterialState.selected);
+    final Color? effectiveActiveColor = widget.activeColor
+      ?? radioThemeData.overlayColor?.resolve(activeStates);
```
Presumably a stateless widget could compute the `Set<MaterialState>` states parameter here based on `selected`.

HansMuller reviewed (commented):
The comments about retaining the stateless widget version of the code should apply to https://github.com/flutter/flutter/pull/97973 and https://github.com/flutter/flutter/pull/97970 as well.

TahaTesser commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -293,58 +296,87 @@ class RadioListTile<T> extends StatelessWidget {
+  State<RadioListTile<T>> createState() => _RadioListTileState<T>();
+}
+
+class _RadioListTileState<T> extends State<RadioListTile<T>> with TickerProviderStateMixin, ToggleableStateMixin {
```
@HansMuller 
states from `states..add(MaterialState.selected);`   is part of the `ToggleableStateMixin` hence I have changed to class to Stateful

HansMuller replied:
https://github.com/flutter/flutter/pull/97972#discussion_r801992173

Why not just construct the states from the widget's selected parameter?

TahaTesser replied:
That makes sense

TahaTesser commented:
@HansMuller 
I updated the code and simplified it as you suggested, this looks cleaner :)

HansMuller commented:
Looking at this again, I'm a bit confused about why we're still computing a default value for Radio.activeColor. If RadioListTile.activeColor is specified we do need to pass it along. If it's not, Radio seems to already do the right thing with respect to RadioTheme and all possible states of the Radio button. In other words, maybe the right thing to do here is to just remove the RadioListTile code that computes `effectiveActiveColor` and just pass along `activeColor`?

TahaTesser commented:
> Looking at this again, I'm a bit confused about why we're still computing a default value for Radio.activeColor. If RadioListTile.activeColor is specified we do need to pass it along. If it's not, Radio seems to already do the right thing with respect to RadioTheme and all possible states of the Radio button. In other words, maybe the right thing to do here is to just remove the RadioListTile code that computes `effectiveActiveColor` and just pass along `activeColor`?

That was unncessary change, only `RadioListTile.activeColor` should be provided to Radio in this class. When  `RadioListTile.activeColor` is non-null, this will be pass to Radio, if not `Radio` already handles its theme

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -326,6 +318,14 @@ class RadioListTile<T> extends StatelessWidget {
+      ?? radioThemeData.overlayColor?.resolve(states)
+      ?? theme.toggleableActiveColor;
     return MergeSemantics(
       child: ListTileTheme.merge(
```
@TahaTesser  - Sorry, I'd made the above comment about 2 weeks ago but had neglected to press the submit button.

TahaTesser replied:
It seems like legacy code, I have removed it from all three widgets :)

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -316,36 +318,42 @@ class RadioListTile<T> extends StatelessWidget {
+    };
+    final Color effectiveActiveColor = activeColor
+      ?? radioThemeData.overlayColor?.resolve(states)
+      ?? theme.toggleableActiveColor;
```
So we're not _replacing_ `toggleableActiveColor` so much as enabling apps to replace it (with RadioTheme). This is a reasonable first step however the next step will be more difficult:

- The default value of `effectiveActiveColor` should be based on a ColorScheme color (not `toggleableActiveColor`).  Some apps currently use `Theme.of(context).toggleableActiveColor` to configure other widgets.  We'll need to replace references to `ThemeData.of(context).toggleableActiveColor` with an expression that looks up the expected ColorScheme color and deals with the difference between `ThemeData.useMaterial3=true` and `ThemeData.useMaterial3=true`.
- Apps that override ThemeData.toggleableActiveColor will need to override ThemeData.radioThemeData instead, and we'll need to create a Dart Fix that makes such a transformation.  It's not clear that this is currently possible - @Piinks what do you think? Something like: 

```dart
ThemeData.light().copyWith(
  toggleableActiveColor: Colors.indigo,
)

// Becomes

ThemeData.light().copyWith(
  radioTheme: RadioThemeData(
    overlayColor: MaterialStateProperty.resolveWith<Color?>((Set<MaterialState> states) {
      return states.contains(MaterialState.selected) ? Colors.indigo : null;
    }),
  ),
)

// Or, if the app has specified other radioTheme properties

ThemeData.light().copyWith(
  toggleableActiveColor: Colors.indigo,
  radioTheme: RadioThemeData(
    splashRadius: 10,
  ),
)

// Becomes

ThemeData.light().copyWith(
  radioTheme: RadioThemeData(
    splashRadius: 10,
    overlayColor: MaterialStateProperty.resolveWith<Color?>((Set<MaterialState> states) {
      return states.contains(MaterialState.selected) ? Colors.indigo : null;
    }),
  ),
)

// Or, worse still

ThemeData.light().copyWith(
  toggleableActiveColor: Colors.indigo,
  radioTheme: MyRadioThemeData(),
)

// Becomes

ThemeData.light().copyWith(
  radioTheme: RadioThemeData(
    splashRadius: 10,
    radioTheme: MyRadioThemeData().copyWith(
      overlayColor: MaterialStateProperty.resolveWith<Color?>((Set<MaterialState> states) {
        return states.contains(MaterialState.selected) ? Colors.indigo : null;
      }),
    ),
  ),
)
```

TahaTesser replied:
@HansMuller 
Thanks for the feedback. I have looked into existing dart fixes and documentation. 

> Apps that override ThemeData.toggleableActiveColor will need to override ThemeData.radioThemeData instead, and we'll need to create a Dart Fix that makes such a transformation. It's not clear that this is currently possible - @Piinks what do you think? Something like:

I was able to replace `toggleableActiveColor` with `RadioListTheme` however, I am not sure if it is possible to check if `RadioListTile` is present in the code or not, since `toggleableActiveColor` is used multiple widgets, the correct theme should be added. 

My original PR https://github.com/flutter/flutter/pull/95870 which led to this PR. It removes `toggleableActiveColor` without adding a theme.

Could we maybe add more details about list tile widgets changes in migration [guide](https://github.com/flutter/website/blob/main/src/release/breaking-changes/toggleable-active-color.md) for the original PR itself instead of adding theme for each widget using dart fix?

TahaTesser replied:
>  if only RadioListTile is present as well as toggleableActiveColor in the same code.

Is there to do this currently in dartfix?

Piinks replied:
No, dart fix cannot do this. That code could be in a separate file or directory, there isn't a way for dart fix to know.

TahaTesser replied:
Hey @HansMuller 

Since `toggleableActiveColor` is also used for widgets other than `RadioListTile`, do you think dart fix to add `RadioListTileData` is still an option here?

Thanks

HansMuller replied:
It sounds like the answer is "no" however in this case and others I think the final default should be a ColorScheme color (presumably ColorScheme.secondary)?  Maybe you're planning to remove this reference in a second step?

HansMuller reviewed (commented):
This looks like a good start at enabling apps to replace `toggleableActiveColor` however we're going to need a plan about how to take the next step, i.e. how to migrate existing code. The alternative is to leave `toggleableActiveColor` intact (as you've done here) rather than deprecating it. Unfortunately, that's unlikely to diminish its use.

Piinks reviewed (commented):
@TahaTesser thanks for your patience. :)

> If other widgets that use toggleableActiveColor also get dart fixes then it will add three themes when running dart fix or we can opt for migration guide without dart fix, please let me know what direction seems best to land this.

There is actually an optional feature for dart fix. This would allow folks to see optional migrations in the IDE and choose one. The command line fix tool does not apply these, since using the `dart fix` command is a bulk operation.

In the case of multiple options for migration, you can just add `bulkApply: false` to the fix rule. That way folks still have guidance, they just have to choose individually. 

Here is one that already exists: 

https://github.com/flutter/flutter/blob/35b18ba2e902dac1e16f557fa385069f1f5e1069/packages/flutter/lib/fix_data.yaml#L503

TahaTesser commented:
Hey @Piinks
Thank you so much for your comment!

> In the case of multiple options for migration, you can just add bulkApply: false to the fix rule. That way folks still have guidance, they just have to choose individually.

I have added other widgets that use `toggleableActiveColor` in the PR and dart fixes for them with `bulkApply: false`. (this will make it easier to review changes or revert if needed since deprecation changes are in this PR too)

There are slight golden changes due to `CheckBox`, `Switch`, `Radio` now using `ColorScheme.secondary`, it is a slightly less blue default color.

Please let me know if I can make any more changes

Piinks commented on dev/integration_tests/flutter_gallery/lib/gallery/themes.dart:
```diff
@@ -30,7 +30,6 @@ ThemeData _buildDarkTheme() {
     primaryColorDark: const Color(0xFF0050a0),
     primaryColorLight: secondaryColor,
     indicatorColor: Colors.white,
-    toggleableActiveColor: const Color(0xFF6997DF),
```
Just a note for follow-up, this is a copy of flutter gallery. The real version in flutter/gallery will need this change as well.

https://github.com/flutter/gallery/blob/637e6a327571da6c3c4011dfa509ecc4377c1ad4/lib/themes/material_demo_theme_data.dart#L19

TahaTesser replied:
That makes sense, i will create PR for that once tests are passing.

Piinks commented on packages/flutter/test_fixes/material.dart.expect:
```diff
@@ -527,4 +527,11 @@ void main() {
 
   // Change made in https://github.com/flutter/flutter/pull/100381
   SelectionOverlay.fadeDuration;
+
```
It looks like this is what is causing test failures here. The test functionality currently only tests the bulk applied fixes, so any `bulkApply: false` ones can't be tested here.

TahaTesser replied:
I've tried both with and without bulk, but tests are still failing. I also separated removed parameter fix, no luck

Piinks replied:
It is not possible right now to test this type of fix.

Piinks replied:
From what I can tell, there should be one dart fix for each option of the migration with bulk apply set to false. If they are not set to bulk apply false, the dart fix tool cannot make the right decision between options and the tests fail.
With bulk apply set to false, nothing should be added to the expect files, because they only test the bulk apply fixes.

Piinks reviewed (commented):
The golden file changes look ok to me, can you update to pass the tests here? I'll run it through google testing after that to check if customers are affected. Thank you!

Piinks commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -800,6 +805,13 @@ class ThemeData with Diagnosticable {
+      'https://flutter.dev/docs/release/breaking-changes/toggleable-active-color#migration-guide. '
+      'This feature was deprecated after v2.12.0-4.2.pre.',
+    )
+    required this.toggleableActiveColor,
```
There is work being done to add a lint that deprecated + required is not allowed. 
Most of these were just updated in https://github.com/flutter/flutter/pull/102107 to use a getter method to avoid a breaking change.

TahaTesser replied:
Thanks! I just made same changes for this properties

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,273 @@
 
 version: 1
 transforms:
+# Changes made in https://github.com/flutter/flutter/pull/95870/
```
These should be as consolidated as possible rather than splitting each change into a separate rule. They should represent  the complete migration paths A, B, C, etc for as many options as there are. `bulkApply: false` should be used since there are options.

TahaTesser replied:
I have consolidated multiple options (based on different conditions) and divided fixes into different theme_data types
(regular, raw, copyWith) and applied bulkApply to false and removed material expect changes

TahaTesser committed:
Migrate `toggleableActiveColor`

Remove `toggleableActiveColor` from gallery

set bulk apply and remove material.expect changes

Update colors and revert a typo

Update fix_data.yaml

Update fix_data for existing themes

.expect tests and minor fixes

Fix conflicts

fix analyzer issue

Update fix_data and .expect test

Update test

Resolve conflict

Update fix

Resolve conflict

TahaTesser commented:
Golden changes are to `Switch`, Checkbox`, `Radio` using ColorScheme color as part of this migration. 
It is not possible to keep  put `toggleableActiveColor` conditionality here since it is not nullable property
```dart
      if (states.contains(MaterialState.selected)) {
        return themeData.colorScheme.secondary;
      }
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,189 @@
 version: 1
 transforms:
+  # Changes made in https://github.com/flutter/flutter/pull/95870/
+  - title: "Migrate 'ThemeData.toggleableActiveColor' to individual themes"
```
What is the difference between the migration options here? This title is shown in the IDE for the user to choose from (In the QuickFix pane), if they all have the same name it won't be clear what migration to pick.

TahaTesser replied:
The difference is different ThemeData constructors 
In this PR, I've divided fixes for `ThemeData`, `ThemeData.copyWith` and `ThemeData.raw`

I can see titles are the same here for different constructors. 

```

  # Changes made in https://github.com/flutter/flutter/pull/41859
  - title: "Remove 'brightness'"
    date: 2020-12-10
    element:
      uris: [ 'cupertino.dart' ]
      constructor: ''
      inClass: 'CupertinoTextThemeData'
    changes:
      - kind: 'removeParameter'
        name: 'brightness'

  # Changes made in https://github.com/flutter/flutter/pull/41859
  - title: "Remove 'brightness'"
    date: 2020-12-10
    element:
      uris: [ 'cupertino.dart' ]
      method: 'copyWith'
      inClass: 'CupertinoTextThemeData'
    changes:
      - kind: 'removeParameter'
        name: 'brightness'
```

Piinks replied:
Ok, I think we have gotten mixed up here somehow. If so I apologize. :) The ones you have referenced are not `bulkApply: false`.

If these are all unique methods, ['', 'raw', 'copyWith'], why are they `bulkApply: false` again? I thought there were more than one choice to migrate each?

For this one in particular ,`constructor: ''`, for the list of `changes`, is the user meant to pick one or apply all of them? Right now, this will apply all of the listed `changes`. If you mean for the user to pick one, they need to each be a different fix rule with `bulkApply: false`. The `title` represents one fix. so if there are optional choices, there should be a separate, unique `title` for each so the user can pick which one they want to apply. This currently will not present the user with choices, but apply all of the listed `changes` when the user picks it.

TahaTesser replied:
That's my bad, I'm fairly new to dart fix. 

I think `bulkApply: false` is not required, I have updated the conditions and removed `bulkApply: false`.

Since this property is used by multiple widgets and we cannot detect which one of them is present in the code hence I am adding all the individual temes like this:

https://user-images.githubusercontent.com/48603081/166324057-3f347bc9-fa9f-412e-99f7-cb2d1ce669a1.mov

TahaTesser replied:
When themes are already present, I am doing .copWith (similar to colorScheme fix) 

Uploading Screen Recording 2022-05-03 at 02.23.05.mov…

Piinks replied:
Not your bad at all, this is a very complicated migration and I think you have provided a very helpful transformation. 

I see now. That is really nice, thank you very much, and apologies again for the mix up! Since they are not `bulkApply`, you can add tests for them in the `.expect` files.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -503,7 +686,7 @@ transforms:
   - title: "Migrate to 'toolbarTextStyle'"
     date: 2021-07-12
-    # TODO(Piinks): Add tests when `bulkApply:false` testing is supported, https://github.com/dart-lang/sdk/issues/44639
+    # TODO(Piinks): Add tests when ` :false` testing is supported, https://github.com/dart-lang/sdk/issues/44639
```
Nit: this can be reverted

Piinks commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -1333,7 +1345,8 @@ class ThemeData with Diagnosticable {
   /// The color used to highlight the active states of toggleable widgets like
   /// [Switch], [Radio], and [Checkbox].
-  final Color toggleableActiveColor;
+  Color get toggleableActiveColor => _toggleableActiveColor!;
```
Nice! 💯

Piinks reviewed (commented):
I'm firing off the internal tests now to see if this breaks anything - thanks for your continued work on this! 
@HansMuller can you take a look? I think this is just about ready.

HansMuller commented on packages/flutter/lib/src/material/switch_list_tile.dart:
```diff
@@ -393,29 +395,35 @@ class SwitchListTile extends StatelessWidget {
+    };
+    final Color effectiveActiveColor = activeColor
+      ?? switchTheme.thumbColor?.resolve(states)
+      ?? theme.toggleableActiveColor;
```
Same comment as earlier. Should this be ColorScheme.secondary?

HansMuller reviewed (commented):
This looks good to me; just some questions about a few remaining references to toggleableActiveColor

devj3ns commented:
Hey guys, do you thing this PR lands before the next stable? Would be awesome if so 😀

Nonetheless great work 💙

TahaTesser commented:
Hey @devj3ns 
Probably not in the next stable but this could be available in a future stable.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,294 @@
+    - kind: 'removeParameter'
+      name: 'toggleableActiveColor'
+    - kind: 'addParameter'
+      index: 95
```
These indices don't seem to match. They should be the index of the parameter in the constructor.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,294 @@
+                     ),"
+        requiredIf: "toggleableActiveColor != '' && checkboxTheme == ''"
+    - kind: 'addParameter'
+      index: 96
```
Like here for example, above there is an option for adding `checkboxTheme` at index 95, and here this option for `checkboxTheme` is index 96?

TahaTesser replied:
Added matching indices

Piinks commented on packages/flutter/test_fixes/material.dart:
```diff
@@ -554,4 +554,11 @@ void main() {
+  ThemeData themeData = ThemeData();
+  themeData = ThemeData(toggleableActiveColor: Colors.black);
+  themeData = ThemeData.raw(toggleableActiveColor: Colors.black);
+  themeData = ThemeData.copyWith(toggleableActiveColor: Colors.black);
```
This is not a static method
```suggestion
  themeData = themeData.copyWith(toggleableActiveColor: Colors.black);
```

TahaTesser replied:
thanks

Piinks commented on packages/flutter/test_fixes/material.dart:
```diff
@@ -554,4 +554,11 @@ void main() {
   TextSelectionOverlay.fadeDuration;
+
+  // Changes made in https://github.com/flutter/flutter/pull/97972
+  ThemeData themeData = ThemeData();
+  themeData = ThemeData(toggleableActiveColor: Colors.black);
+  themeData = ThemeData.raw(toggleableActiveColor: Colors.black);
+  themeData = ThemeData.copyWith(toggleableActiveColor: Colors.black);
```
```suggestion
  ThemeData themeData = ThemeData();
  themeData = ThemeData(toggleableActiveColor: Colors.black);
  themeData = ThemeData(
    toggleableActiveColor: Colors.black,
  );
  themeData = ThemeData.raw(toggleableActiveColor: Colors.black);
  themeData = ThemeData.raw(
    toggleableActiveColor: Colors.black,
  );
  themeData = themeData.copyWith(toggleableActiveColor: Colors.black);
  themeData = themeData.copyWith(
    toggleableActiveColor: Colors.black,
  );
```

Can you add these test cases? The failing customer test is showing that too many commas are being inserted when the dart fix is applied. It appears to be based on formatting - one line or multiline.

 I can't tell if it is a bug with the fix, or a bug with the fix tool itself.

TahaTesser replied:
Just added :)

Piinks commented on packages/flutter/test_fixes/material.dart:
```diff
@@ -554,4 +554,20 @@ void main() {
+  themeData = ThemeData(
+    toggleableActiveColor: Colors.black,
+  );
+  themeData = themeData.raw(toggleableActiveColor: Colors.black);
+  themeData = themeData.raw(
```
This is a constructor, in contrast to `copyWith`, which is a method.

```suggestion
  themeData = ThemeData.raw(toggleableActiveColor: Colors.black);
  themeData = ThemeData.raw(
```

Analysis is disabled for this file because it is intentionally full of broken code. It may be helpful to put the 'before' code under normal analysis to catch typos like this.

TahaTesser replied:
That's my bad,  I didn't realize it, thank you

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,294 @@
+    - kind: 'removeParameter'
+      name: 'toggleableActiveColor'
+    - kind: 'addParameter'
+      index: 96
```
Why is this `checkboxTheme` 96, and the one below `97`? Shouldn't they be the same since they represent the same parameter? I would check all of these indices. Since this adds a lot of code, we should make sure it is getting inserted in the right place.

TahaTesser replied:
This is `checkboxTheme()` and below is the `checkboxTheme.copyWith`. If I add matching indices for both, dart fix throws `Error from the analysis server: Exception while getting bulk fixes:`.

Piinks commented:
I think there is a dart fix issue, I have filed https://github.com/dart-lang/sdk/issues/49065
I am also migrating the customer test that is blocking this due to the issue: https://github.com/flutter/gallery/pull/694

Piinks commented:
When https://github.com/flutter/tests/pull/133 lands, the customer test failure here should be unblocked.

Piinks reviewed (approved):
LGTM! It looks like there are some merge conflicts though. I will run this through internal testing in the mean time to see that it is not breaking.

Piinks reviewed (commented):
This breaks a bunch of tests internally. Is there anyway to deprecate it while maintaining backwards compatibility during the deprecation period? That is how we usually give folks time to migrate.

TahaTesser commented:
> This breaks a bunch of tests internally. Is there anyway to deprecate it while maintaining backwards compatibility during the deprecation period? That is how we usually give folks time to migrate.

`toggleableActiveColor` is non-nullable so it cannot fallback to `ColorScheme` color.

I think a better solution would be to add the `useMaterial3` condition (we did the same here to upgrade list_tile font https://github.com/flutter/flutter/pull/102167 which was reverted earlier without this condition).


So in a nutshell, when using Material 3, these widgets won't utilize the `Theme.toggleableActiveColor` property.
Please let me know if this sounds reasonable.

TahaTesser commented:
If we use material 3 condition, it would also fix https://github.com/flutter/flutter/issues/104446

Piinks commented:
Hmm, that does not sound super clear as a user, that choosing Material 3 makes this change. The more changes we put behind the Material 3 flag, the more difficult it will be to make it the new default.

It also is not a sure thing that going with that approach won't still break internal tests. There are a lot that have already adopted Material 3. Let me see what I can do to get customers migrated, I'll sync up with @HansMuller and @darrenaustin about the M3 flag as well.

Piinks reviewed (changes requested):
Working on migrating customers, I found that the migration you have added here does not actually work as expected. It does not account for disabled state, and if you look in the switch implementation, an opacity is applied to the switch track based on the thumb color.

It looks like it really should be something like this:

```dart
ThemeData(
...
checkboxTheme: CheckboxThemeData(
    fillColor: MaterialStateProperty.resolveWith((states) {
      if (states.contains(MaterialState.disabled)) {
        return null;
      }
      if (states.contains(MaterialState.selected)) {
        return myMigratedColor;
      }
      return null;
    }),
  ),
SwitchThemeData(
    thumbColor: MaterialStateProperty.resolveWith((Set<MaterialState> states) {
      if (states.contains(MaterialState.disabled)) {
        return null;
      }
      if (states.contains(MaterialState.selected)) {
        return myMigratedColor;
      }
      return null;
    }),
  ),
radioTheme: RadioThemeData(
    fillColor: MaterialStateProperty.resolveWith((states) {
      if (states.contains(MaterialState.disabled)) {
        return null;
      }
      if (states.contains(MaterialState.selected)) {
        return myMigratedColor;
      }
      return null;
    }),
  ),
...
);
```

Can you fix this?

Piinks commented:
I saw https://github.com/flutter/website/pull/6626, usually we land migration guides after the change lands, right now this doc is not true. It should be updated to include these component theme migrations as well. That's why we land the guide last, migrating customers usually reveals what guidance is needed.

Piinks commented:
I am back from being off for a few days, I will check on the status of the remaining customers and report back here tomorrow.

TahaTesser commented:
Working on updating the migration guide.

TahaTesser commented:
> I saw https://github.com/flutter/website/pull/6626, usually we land migration guides after the change lands, right now this doc is not true. It should be updated to include these component theme migrations as well. That's why we land the guide last, migrating customers usually reveals what guidance is needed.

That makes sense, I'll keep that in mind for future PRs.

Filed https://github.com/flutter/website/pull/7243, I'm not sure if the wording needs update too.

Piinks commented:
This one as well. @TahaTesser can you update this branch with the latest from master? I am having trouble using our tools to sync this with the internal copy for testing. Thanks!

Piinks commented:
This one is very close to ready with customers, I am going to focus on completing this one before finishing the migration for https://github.com/flutter/flutter/pull/99753. It will take longer for both if I try to do them simultaneously.

Piinks commented:
@TahaTesser following up from https://github.com/flutter/website/pull/7243 it looks like there may be a bug here, but I am not sure..

Using the default ThemeData() constructor will result in default values for toggleableActiveColor:
light mode: ThemeData.accentColor ?? ThemeData.primarySwatch[600]
dark mode: Colors.tealAccent[200]

and ThemeData.primarySwatch would default to Colors.blue if null.

It looks like this constructor does not set the ColorScheme.secondary when a colorScheme is not provided. Since ColorScheme.secondary is the new default for radio buttons, should it reflect these colors in this case? Or does it already? I found this while migrating a customer.

TahaTesser commented:
@Piinks 
Really appreciate you looking into this, I'll deep dive and try to figure out what's going on.

Piinks commented:
Of course! No worries - and I could be wrong. Migrating theme info for customers is always tricky. They could have multiple themes defined anywhere in their app, so it could be that there is no bug here. It's just worth double checking on a breaking change.

TahaTesser commented:
Hey @Piinks 

> Using the default ThemeData() constructor will result in default values for toggleableActiveColor:
> light mode: ThemeData.accentColor ?? ThemeData.primarySwatch[600]
> dark mode: Colors.tealAccent[200]
> 
> and ThemeData.primarySwatch would default to Colors.blue if null.

I looked into this.

If you look at the specs https://material.io/components/switches#theming, ColorScheme secondary is the only color to be used. Colors defined in `toggleableActiveColor` are hard coded. Looks like it was created when `ThemeData.primarySwatch` was more prominent part of the theme system, now it is largely ignored.

IMO, this migration doesn't need to be compatible with `ThemeData.primarySwatch` since it is also need to be [deprecated](https://github.com/flutter/flutter/issues/91772). 
[theme_data largely ignores primarySwatch when brightness is set to dark](https://github.com/flutter/flutter/issues/19089) and Darren Austin's [comment](https://github.com/flutter/flutter/issues/19089#issuecomment-1104801534)  also indicates `ColorScheme` colors should be default (matching Material specs here too).


> In this case they are broken because ColorScheme.secondary is a different default.

IMO, `ColorScheme.secondary` is slightly lighter blue by default hence the difference (also native switch is ligher too compared to the one created using `toggleableActiveColor but it is a spec accurate color regardless of the theme brightness (ColorScheme.secondary). In fact track color is even lighter in native Android.

| Flutter | Android |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/173349530-b9b1d3fd-70e7-473d-8060-dbcba1e96ee8.png" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/173348725-bfc07696-454a-4400-87f4-35fb1ef1d4f7.png" height="450" /> |



>  Since ColorScheme.secondary is the new default for radio buttons, should it reflect these colors in this case?

I think widgets should reflect `ColorScheme.secondary`.

> It looks like this constructor does not set the ColorScheme.secondary when a colorScheme is not provided.


Default `ColorScheme.secondary` color  is returned when no colorScheme is provided. Widgets are still blue when you don't define `ColorScheme`.

```dart
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Example(),
    );
```



In cases where  `ColorScheme.secondary` is already defined but it is a different color than what is defined in `ThemeData.toggleableActiveColor`, app should not break since fix overrides the component themes to match existiing `toggleableActiveColor`, if component is already defined, fix applies `.copyWith` and adds `toggleableActiveColor` color for each widget.

> It's just worth double checking on a breaking change.

Agreed. Without looking at specific customer tests, i cannot say if this is deal breaker and how to overcome this if needed but looking the specs and deprecate primarySwatch issue. I cannot find anything else to fix here.

TahaTesser commented:
I'm back long weekend break and I see some conflict in this PR. Updating for conflicts and please let me if my feedback https://github.com/flutter/flutter/pull/97972#issuecomment-1153834494 makes sense and what can we do to land this.

Piinks commented:
Welcome back! Thank you. I apologize for this taking so long. Most of the time, we'll deprecate something without breaking, and customers migrate on their own time. Since this is a hard break, we have to fix them all first and it is a slow process. Thanks for your patience!

Piinks commented:
If you want, you can likely hold off on resolving merge conflicts until it is ready.

Piinks commented:
@TahaTesser this is finally ready! Can you update once more so we can get this ready to finally land! :)

TahaTesser commented:
@Piinks 
I resolve the conflicts and rebased the branch. 
However test fixes passes locally 
```
tahatesser@Tahas-MacBook-Pro test_fixes % dart fix --compare-to-golden
Computing fixes in test_fixes... 32.3s
Passed: 6, Failed: 0
tahatesser@Tahas-MacBook-Pro test_fixes % 
```

It doesn't seem to pass in Flutter CI

Piinks commented:
How odd! I will check this morning and see if I can resolve them.

Piinks commented:
QQ - did you merge or rebase?

Piinks commented:
It looks like the only difference between Expected/Actual in CI is white space. I thought dart fix was supposed to ignore whitespace.. let me inquire with the dart team.

TahaTesser commented:
>  QQ - did you merge or rebase?

Resolved conflicts and rebased.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     )"
```
See here, no comma. The others should reflect the same.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

TahaTesser commented:
> This PR is pretty old so you may have something cached locally that is out of sync with CI.

Let me try a different machine, where I recently Installed Flutter on it.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,438 @@
+                         }\n
+                         return null;\n
+                       }),\n
+                     ),"
```
```suggestion
                     )"
```

Piinks reviewed (commented):
I will commit these and check CI

Piinks reviewed (commented):
Oh, I do not have permission on this PR to modify it. Can you give permission or commit these changes below? I'd like to get this landed quickly. Customer code is constantly changing internally and it could be blocked again.

TahaTesser commented:
I can confirm it's the cache issue. The test fails on a different machine locally. 

>  Can you give permission or commit these changes below? 

I will make the changes.

Piinks commented:
Fab! Thank you for being so attentive to this. We will both celebrate this landing for sure!

TahaTesser committed:
Update commas

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)
Hold for final checks, I am running this through internal tests one. last. time. 🍾

Piinks commented:
Is it passing locally for you again? It is still failing here, although differently. Checking the diff now.

Piinks commented on packages/flutter/test_fixes/material.dart.expect:
```diff
@@ -538,4 +538,254 @@ void main() {
+       }
+ 	     return null;
+ 	   }),
+ 	  )switchTheme: SwitchThemeData(
```
Several of your expectations here are missing commas now.. why?

```suggestion
 	  ), switchTheme: SwitchThemeData(
```

Piinks replied:
In case this was just a wrong-file change, the commas should removed from the fix_data.yaml. Not the .expect files.

TahaTesser replied:
> In case this was just a wrong-file change, the commas should removed from the fix_data.yaml. Not the .expect files.

The test fails as .expect doesn't generate commas on my machine after updating `fix_data.yaml`.

![Screenshot 2022-08-02 at 19 56 44](https://user-images.githubusercontent.com/48603081/182431665-7befa632-53ec-417c-912e-ccea40a840ef.png)

TahaTesser replied:
I also deleted `bin/cache` folder before testing.

TahaTesser replied:
I am updating this again on a different machine, I can see commas are generated again. Apologies, I have to reset my main machine.

TahaTesser committed:
fix commas

TahaTesser commented:
@Piinks 
Finally green :)

Piinks commented:
Hooray! Internal tests are running right now one last time, and then we can merge
